plaidml/core/core.h

```diff
@@ -5,12 +5,32 @@
 #include <cstring>
 #include <functional>
 #include <memory>
+#include <sstream>
 #include <stdexcept>
 #include <string>
 #include <vector>
 
 #include "plaidml/core/ffi.h"
 
+#if __has_include(<source_location>)
```

Comment:
You mentioned that this only works on C++17 and beyond, I think that's going to be a blocker. I think we need to support C++11 and beyond.

Reply:
I am fairly confident that if `__has_include()` isn't defined, it will default to the `#else` block automatically. I tested that with a `#if nonsensefunction(morenonsense)` which worked as desired. If you think I got this functionality wrong lmk and I can look at alternatives.

Reply:
Instead of being "fairly confident," would it make sense to just test this on c++11 and see what happens?
In my mind this could be something as simple as building the target (and perhaps a small unit test) with `bazel build --cxxopt='-std=c++11'` but perhaps there are other dependencies I'm missing; perhaps Frank has a better idea.

Reply:
I was unable to build with c++11 due to llvm (and coincidentally, my code, for the same reason): both use `enable_if_t` which is not available in c++11. I believe the fix to that for my code would be a simple as changing it to `enable_if<..., true>`, but can't easily verify that works since nothing depending on llvm will build for me if I specify `--cxxopt='-std=c++11'`

Reply:
However `--cxxopt='-std=c++14'` builds & runs fine which further verifies my usage of `__has_include()` since it shouldn't be defined for c++14

Reply:
Right, so in this situation, we need to figure out how to build just the edsl public portion with C++11 since LLVM does require C++14. I don't have a good solution to offer right now, it might require some tricky CMake code. It might be good enough to test with C++14 for now. Also @dgkutnic can we get some requirements on whether C++14 would be acceptable?

Reply:
@flaub from previous discussions I've had w/ others, my impression was that C++11 and above was the requirement. However, I'll take the AR to cycle back on this and get clarity for what we need for a minimum viable product. Ideally, we could get this landed with C++14 and above for the initial pass, and work later on the C++11 backwards compatibility when necessary. Happy to take this conversation offline.

Reply:
I tested C++11 with `source_location` out locally outside of `edsl.h` with this:
```
#include <iostream>

#if __has_include(<source_location>)
#include <source_location>
using edsl_source_location = std::source_location;
#elif __cplusplus>=201402 and __has_include(<experimental/source_location>)
#include <experimental/source_location>
using edsl_source_location = std::experimental::source_location;
#else
struct edsl_source_location {
  static edsl_source_location current() { return edsl_source_location{}; }
  constexpr uint_least32_t line() const noexcept { return 0; }
  constexpr uint_least32_t column() const noexcept { return 0; }
  constexpr const char* file_name() const noexcept { return "??"; }
  constexpr const char* function_name() const noexcept { return "??"; }
};
#endif

int main() {
    auto loc = edsl_source_location::current();
    std::cout<<loc.file_name()<<":"<<loc.line()<<std::endl;
}
```

There was an issue with `__has_include`, but not the one expected; `__has_include` itself actually works fine back through C++11 (ie, regardless of C++ version `edsl_source_location` always gets defined). However, apparently with C++11 some experimental version of the `source_location` header is found, but not one which compiles (this appears to be a [known issue](https://features.isocpp.open-std.narkive.com/RkxjCiAU/sg10-checking-has-include-on-its-own-is-not-sufficient) with `__has_include`). The additional check to ensure C++>=14 handles this. So, I have vetted the definition of `source_location` for C++11-17 locally. 

As for the `enable_if` functionality that allows `source_location` to be passed through operator overloads, I switched from `enable_if_t` (which was only added in C++14) to `enable_if` (which should be fine for C++11). I've checked this for C++14-17, however I can't do a meaningful check for C++11 since llvm does not compile with C++11.

TLDR, the code I'm introducing _should_ work for C++11-17; it's been thoroughly tested for C++14-17; it's been tested to the extent possible for C++11 (barring implementing the tricky CMake code Frank mentioned).

Reply:
SGTM